streaming_rows.go

```diff
@@ -0,0 +1,348 @@
+		}
+	}
+	h := http.Header{}
+	if sessionID != nil {
+		h.Add("ds-session-id", *sessionID)
+	}
+
+	conn, resp, err := dialer.DialContext(ctx, u.String(), h)
+	if err != nil {
+		if resp != nil && resp.StatusCode != 200 {
+			b, err := io.ReadAll(resp.Body)
+			if err != nil {
+				return nil, &ErrClientError{message: "unable to read dataplane response", wrapErr: err}
+			}
+			return nil, &ErrInterfaceError{message: string(b)}
+		}
+		return nil, err
+	}
+
+	if err = conn.WriteJSON(&AuthMessage{
+		Type:        "auth",
+		AccessToken: req.Token,
+		SessionID:   ptr.Deref(sessionID, ""),
+	}); err != nil {
+		return nil, &ErrInterfaceError{message: "unable to send request", wrapErr: err}
+	}
+
+	rows := &streamingRows{
+		conn:      conn,
+		dataChan:  make(chan *PrintTopicDataMessage, 30),
+		readyChan: make(chan struct{}),
+	}
+	go rows.readMessages()
+	<-rows.readyChan
+
+	return rows, nil
+}
+
+func (r *streamingRows) readMessages() {
+	r.conn.SetReadDeadline(time.Time{})
+	for {
+		var msg PrintTopicMessage
+		if err := r.conn.ReadJSON(&msg); err != nil {
+			r.connErr = &ErrInterfaceError{message: "unable to read message from server", wrapErr: err}
+			return
+		}
+		switch msg.Type {
+		case "error":
+			r.connErr = &ErrSQLError{SQLCode: msg.Err.SqlCode, Message: msg.Err.Message}
+			close(r.readyChan)
+			return
+		case "metadata":
+			r.metadata = &msg.Metadata
+			close(r.readyChan)
+		case "data":
+			r.dataChan <- &msg.Data
+		default:
+			r.connErr = &ErrInterfaceError{message: "unexpected message type " + msg.Type}
+			close(r.readyChan)
+			return
+		}
+	}
+}
+
+func (r *streamingRows) ColumnTypeNullable(index int) (nullable bool, ok bool) {
+	if r.metadata == nil {
+		return false, false
+	}
+	if index < 0 || index >= len(r.metadata.Columns) {
+		return false, false
+	}
+	return r.metadata.Columns[index].Nullable, true
+}
+
+func (r *streamingRows) ColumnTypeDatabaseTypeName(index int) string {
+	if r.metadata == nil {
+		return ""
+	}
+	if index < 0 || index >= len(r.metadata.Columns) {
+		return ""
+	}
+	return r.metadata.Columns[index].Name
+}
+
+func (r *streamingRows) ColumnTypeScanType(index int) reflect.Type {
+	if r.metadata == nil {
+		return nil
+	}
+	if index < 0 || index >= len(r.metadata.Columns) {
+		return nil
+	}
+	md := r.metadata.Columns[index]
+	switch {
+	case strings.HasPrefix(md.Type, "VARCHAR"):
+		return typeMap["VARCHAR"]
+	case strings.HasPrefix(md.Type, "DECIMAL"):
+		return typeMap["DECIMAL"]
+	case strings.HasPrefix(md.Type, "TIMESTAMP"):
+		return typeMap["TIMESTAMP"]
+	case strings.HasPrefix(md.Type, "TIME"):
+		return typeMap["TIME"]
+	case strings.HasPrefix(md.Type, "ARRAY"):
+		return typeMap["ARRAY"]
+	case strings.HasPrefix(md.Type, "STRUCT"):
+		return typeMap["STRUCT"]
+	case strings.HasPrefix(md.Type, "MAP"):
+		return typeMap["MAP"]
+	default:
+		return typeMap[md.Type]
+	}
+}
+
+func (r *streamingRows) Close() error {
+	r.metadata = nil
+	close(r.dataChan)
```

Comment:
is close(r.readyChan) needed?

Reply:
readyChan may already be closed, closing it again will cause a panic.